### PR TITLE
Cleanup/precondition error msgs

### PR DESCRIPTION
**Pull request by christianhoff**
**Motivation**
Precondition checks in builder methods must have meaningful error messages.

**Changes**
* add error messages to precondition checks
* refactor TopicAliasUsage enum to only have states YES, NO and IF_AVAILABLE
* remove redundant precondition check on toipc alias usage in publish builder

### CODE REVIEW

File: src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5WillPublishBuilder.java [hunk 1]
```diff
@@ -199,19 +199,23 @@ public Mqtt5WillPublishBuilder(@Nullable final Function<? super Mqtt5Publish, P>
     }
 
     @NotNull
-    public Mqtt5WillPublishBuilder<P> delayInterval(final long delayInterval) {
-        Preconditions.checkArgument(UnsignedDataTypes.isUnsignedInt(delayInterval));
-        this.delayInterval = delayInterval;
+    public Mqtt5WillPublishBuilder<P> delayInterval(final long delayInterval, @NotNull final TimeUnit timeUnit) {
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5WillPublishBuilder.java]: 👍

File: src/main/java/org/mqttbee/mqtt/util/MqttBuilderUtil.java [hunk 2]
```diff
@@ -97,16 +97,16 @@ public static MqttTopicFilterImpl topicFilter(@NotNull final String string) {
 
     @NotNull
     public static MqttTopicFilterImpl topicFilter(@NotNull final MqttTopicFilter topic) {
-        Preconditions.checkNotNull(topic);
+        Preconditions.checkNotNull(topic, "Topic must not be null.");
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/util/MqttBuilderUtil.java]: Topic Filter

Comment (SgtSilvio) [reply src/main/java/org/mqttbee/mqtt/util/MqttBuilderUtil.java]: you can also rename the argument

File: src/main/java/org/mqttbee/mqtt/util/MqttBuilderUtil.java [hunk 3]
```diff
@@ -118,7 +118,7 @@ public static MqttSharedTopicFilterImpl sharedTopicFilter(
 
     @NotNull
     public static MqttClientIdentifierImpl clientIdentifier(@NotNull final String string) {
-        Preconditions.checkNotNull(string);
+        Preconditions.checkNotNull(string, "string must not be null.");
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/util/MqttBuilderUtil.java]: **S**tring to be consistent

File: src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java [hunk 4]
```diff
@@ -45,14 +46,14 @@ public MqttTopicAliasMapping(final int topicAliasMaximum) {
...
         int topicAlias = MqttStatefulPublish.DEFAULT_NO_TOPIC_ALIAS;
-        if (topicAliasUsage != TopicAliasUsage.MUST_NOT) {
-            if (nextTopicAlias > topicAliasMaximum) {
-                if (topicAliasUsage == TopicAliasUsage.MAY_OVERWRITE) {
-                    topicAlias = random.nextInt(topicAliasMaximum) + 1;
-                    hashMap.values().remove(topicAlias);
-                    hashMap.put(topic.toString(), topicAlias);
-                }
-            } else {
+        if (!unusedTopicAliasAvailable()) {
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java]: If we are changing this now, I think the following is even simpler:
```
final int topicAlias = DEFAULT_NO_TOPIC_ALIAS;
if ((topicAliasUsage == YES) || (topicAliasUsage == IF_AVAILABLE)) {
    if (nextTopicAlias <= topicAliasMaximum) {
        topicAlias = nextTopicAlias++;
        hashMap.put(topic.toString(), topicAlias);
    } else if (topicAliasUsage == YES) {
        topicAlias = random.nextInt(topicAliasMaximum) + 1;
        hashMap.values().remove(topicAlias);
        hashMap.put(topic.toString(), topicAlias);
    }
}
return topicAlias;
```
I think it is easier to understand if we first handle the normal case (nextTopicAlias <= topicAliasMaximum) and then the overwriting. With this I think we can close the old PR #25, as this is very concise.

File: src/main/java/org/mqttbee/mqtt/util/MqttBuilderUtil.java (hunk 2 above)
Comment (christianhoff) [reply src/main/java/org/mqttbee/mqtt/util/MqttBuilderUtil.java]: DONE

File: src/main/java/org/mqttbee/mqtt/util/MqttBuilderUtil.java (hunk 3 above)
Comment (christianhoff) [reply src/main/java/org/mqttbee/mqtt/util/MqttBuilderUtil.java]: DONE

File: src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java (hunk 4 above)
Comment (christianhoff) [reply src/main/java/org/mqttbee/mqtt/message/publish/MqttTopicAliasMapping.java]: DONE

File: src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5PublishBuilder.java [hunk 5]
```diff
@@ -58,7 +57,7 @@
     MqttUTF8StringImpl contentType;
     MqttTopicImpl responseTopic;
     ByteBuffer correlationData;
-    private TopicAliasUsage topicAliasUsage = DEFAULT_TOPIC_ALIAS_USAGE;
+    private TopicAliasUsage useTopicAlias = DEFAULT_TOPIC_ALIAS_USAGE;
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5PublishBuilder.java]: Minor: the attribute and parameter should stay topicAliasUsage

File: src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5PublishBuilder.java [hunk 6]
```diff
@@ -182,10 +183,8 @@ public Mqtt5PublishBuilder(@Nullable final Function<? super Mqtt5Publish, P> par
     }
 
     @NotNull
-    public Mqtt5PublishBuilder<P> topicAliasUsage(@NotNull final TopicAliasUsage topicAliasUsage) {
-        Preconditions.checkArgument(
-                topicAliasUsage == MUST_NOT || topicAliasUsage == MAY || topicAliasUsage == MAY_OVERWRITE);
-        this.topicAliasUsage = topicAliasUsage;
+    public Mqtt5PublishBuilder<P> useTopicAlias(@NotNull final TopicAliasUsage useTopicAlias) {
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5PublishBuilder.java]: Minor: the attribute and parameter should stay topicAliasUsage

File: src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5WillPublishBuilder.java [hunk 7]
```diff
@@ -181,7 +181,7 @@ public Mqtt5WillPublishBuilder(@Nullable final Function<? super Mqtt5Publish, P>
     @NotNull
     @Override
     @Deprecated
-    public Mqtt5WillPublishBuilder<P> topicAliasUsage(@NotNull final TopicAliasUsage topicAliasUsage) {
+    public Mqtt5WillPublishBuilder<P> useTopicAlias(@NotNull final TopicAliasUsage useTopicAlias) {
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5WillPublishBuilder.java]: Minor: the attribute and parameter should stay topicAliasUsage

File: src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/TopicAliasUsage.java [hunk 8]
```diff
@@ -21,29 +21,27 @@
...
      */
-    MAY,
+    NO,
     /**
-     * Indicates that an outgoing PUBLISH packet may use a topic alias and also may overwrite an existing topic alias
-     * mapping.
+     * Indicates that an outgoing PUBLISH packet will use a topic alias, if there are some unused topic aliases
+     * available.
+     *
+     * @see org.mqttbee.api.mqtt.mqtt5.message.connect.Mqtt5ConnectRestrictionsBuilder#topicAliasMaximum(int)
```
Comment (SgtSilvio) [inline src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/TopicAliasUsage.java]: The reference should be for Mqtt5ConnAckRestrictions#getTopicAliasMaximum() (Mqtt5ConnectRestrictionsBuilder#topicAliasMaximum(int) is for incoming topic aliases)

File: src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5PublishBuilder.java (hunk 5 above)
Comment (christianhoff) [reply src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5PublishBuilder.java]: DONE

File: src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5PublishBuilder.java (hunk 6 above)
Comment (christianhoff) [reply src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5PublishBuilder.java]: DONE

File: src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5WillPublishBuilder.java (hunk 7 above)
Comment (christianhoff) [reply src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/Mqtt5WillPublishBuilder.java]: DONE

File: src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/TopicAliasUsage.java (hunk 8 above)
Comment (christianhoff) [reply src/main/java/org/mqttbee/api/mqtt/mqtt5/message/publish/TopicAliasUsage.java]: DONE

**Pull request merged**